vowpalwabbit/reductions/epsilon_decay.cc

```diff
@@ -0,0 +1,240 @@
+      reset_models(n_iter, end_iter, data.weights, data.epsilon_decay_alpha, data.epsilon_decay_tau, model_count);
+      break;
+    }
+  }
+}
+
+void save_load_epsilon_decay(epsilon_decay_data& epsilon_decay, io_buf& io, bool read, bool text)
+{
+  if (io.num_files() == 0) { return; }
+  if (read) { VW::model_utils::read_model_field(io, epsilon_decay); }
+  else
+  {
+    VW::model_utils::write_model_field(io, epsilon_decay, "_epsilon_decay", text);
+  }
+}
+
+VW::LEARNER::base_learner* epsilon_decay_setup(VW::setup_base_i& stack_builder)
+{
+  VW::config::options_i& options = *stack_builder.get_options();
+  VW::workspace& all = *stack_builder.get_all_pointer();
+
+  std::string arg;
+  bool epsilon_decay_option;
+  uint64_t model_count;
+  uint64_t min_scope;
+  float epsilon_decay_alpha;
+  float epsilon_decay_tau;
+
+  option_group_definition new_options("Epsilon-Decaying Exploration");
+  new_options
+      .add(make_option("epsilon_decay", epsilon_decay_option)
+               .necessary()
+               .keep()
+               .help("Use decay of exploration reduction"))
+      .add(make_option("model_count", model_count).keep().default_value(3).help("Set number of exploration models"))
+      .add(make_option("min_scope", min_scope)
+               .keep()
+               .default_value(100)
+               .help("Minimum example count of model before removing"))
+      .add(make_option("epsilon_decay_alpha", epsilon_decay_alpha)
+               .keep()
+               .default_value(DEFAULT_ALPHA)
+               .help("Set confidence interval for champion change"))
+      .add(make_option("epsilon_decay_tau", epsilon_decay_tau)
+               .keep()
+               .default_value(DEFAULT_TAU)
+               .help("Time constant for count decay"));
+
+  if (!options.add_parse_and_check_necessary(new_options)) { return nullptr; }
+
+  // Update model count to be 2^n
+  uint64_t ppw = 1;
+  while (ppw < model_count) { ppw *= 2; }
+
+  auto data =
+      VW::make_unique<epsilon_decay_data>(model_count, min_scope, epsilon_decay_alpha, epsilon_decay_tau, all.weights);
+
+  // make sure we setup the rest of the stack with cleared interactions
+  // to make sure there are not subtle bugs
+  auto* base_learner = stack_builder.setup_base_learner();
+  if (base_learner->is_multiline())
+  {
+    auto* learner = VW::LEARNER::make_reduction_learner(std::move(data), VW::LEARNER::as_multiline(base_learner), learn,
+        predict, stack_builder.get_setupfn_name(epsilon_decay_setup))
+                        .set_params_per_weight(ppw)
+                        .set_output_prediction_type(base_learner->get_output_prediction_type())
+                        .set_save_load(save_load_epsilon_decay)
+                        .build();
+
+    return VW::LEARNER::make_base(*learner);
+  }
+  else
+  {
+    // not implemented yet
+    THROW("Epsilon decay does not supported for single line learners yet");
```

Comment:
I would avoid saying "yet"

```suggestion
    THROW("Epsilon decay  (--epsilon_decay) does not supported for single line learners");
```